Adding dequantize_val and requantize_val

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23909 Adding dequantize_val and requantize_val**

Differential Revision: [D16678276](https://our.internmc.facebook.com/intern/diff/D16678276)

```
$> ./build/bin/quantized_test
Running main() from ../third_party/googletest/googletest/src/gtest_main.cc
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from TestQTensor
[ RUN      ] TestQTensor.QuantDequantAPIs
[       OK ] TestQTensor.QuantDequantAPIs (0 ms)
[ RUN      ] TestQTensor.RoundingMode
[       OK ] TestQTensor.RoundingMode (0 ms)
[ RUN      ] TestQTensor.Item
[       OK ] TestQTensor.Item (0 ms)
[ RUN      ] TestQTensor.EmptyQuantized
[       OK ] TestQTensor.EmptyQuantized (0 ms)
[----------] 4 tests from TestQTensor (0 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (0 ms total)
[  PASSED  ] 4 tests.
```

Commit:
Adding dequantize_val and requantize_val

Commit:
Update on "Adding dequantize_val and requantize_val"

Commit:
Update on "Adding dequantize_val and requantize_val"


Differential Revision: [D16678276](https://our.internmc.facebook.com/intern/diff/D16678276)

jerryzh168 commented:
should we add some tests in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/quantized_test.cpp

Commit:
Update on "Adding dequantize_val and requantize_val"


Differential Revision: [D16678276](https://our.internmc.facebook.com/intern/diff/D16678276)

jerryzh168 commented on aten/src/ATen/test/quantized_test.cpp:
```diff
@@ -45,6 +46,24 @@ TEST(TestQTensor, QuantDequantAPIs) {
+  Tensor reqr = at::quantize_linear(r, new_scale, new_zero_point, kQInt8);
+  auto reqr_data = reqr.data<qint8>();
+  for (auto i = 0; i < num_elements; ++i) {
+    reqr_data[i].val_ = requantize_val<quint8, qint8>(scale, zero_point,
```
can you fix the formatting?

zafartahirov replied:
Can you point to the actual formatting issue, please?

jerryzh168 replied:
looks like the arguments are expanded into multiple lines and doesn't align. not sure whether this is just github display it this way or not. just make sure it looks ok in your editor

zafartahirov replied:
Resolved in person

jerryzh168 commented on aten/src/ATen/test/quantized_test.cpp:
```diff
@@ -82,7 +101,9 @@ TEST(TestQTensor, EmptyQuantized) {
   int val = 100;
   int numel = 10;
-  Tensor q = at::_empty_affine_quantized({numel}, at::device(at::kCPU).dtype(kQUInt8), scale, zero_point);
+  Tensor q = at::_empty_affine_quantized({numel},
```
same here

Commit:
Update on "Adding dequantize_val and requantize_val"


Differential Revision: [D16678276](https://our.internmc.facebook.com/intern/diff/D16678276)

```
$> ./build/bin/quantized_test
Running main() from ../third_party/googletest/googletest/src/gtest_main.cc
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from TestQTensor
[ RUN      ] TestQTensor.QuantDequantAPIs
[       OK ] TestQTensor.QuantDequantAPIs (0 ms)
[ RUN      ] TestQTensor.RoundingMode
[       OK ] TestQTensor.RoundingMode (0 ms)
[ RUN      ] TestQTensor.Item
[       OK ] TestQTensor.Item (0 ms)
[ RUN      ] TestQTensor.EmptyQuantized
[       OK ] TestQTensor.EmptyQuantized (0 ms)
[----------] 4 tests from TestQTensor (0 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (0 ms total)
[  PASSED  ] 4 tests.
```

Commit:
Update on "Adding dequantize_val and requantize_val"


Differential Revision: [D16678276](https://our.internmc.facebook.com/intern/diff/D16678276)

```
$> ./build/bin/quantized_test
Running main() from ../third_party/googletest/googletest/src/gtest_main.cc
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from TestQTensor
[ RUN      ] TestQTensor.QuantDequantAPIs
[       OK ] TestQTensor.QuantDequantAPIs (0 ms)
[ RUN      ] TestQTensor.RoundingMode
[       OK ] TestQTensor.RoundingMode (0 ms)
[ RUN      ] TestQTensor.Item
[       OK ] TestQTensor.Item (0 ms)
[ RUN      ] TestQTensor.EmptyQuantized
[       OK ] TestQTensor.EmptyQuantized (0 ms)
[----------] 4 tests from TestQTensor (0 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (0 ms total)
[  PASSED  ] 4 tests.
```

jerryzh168 reviewed (approved):
Thanks!

Commit:
Update on "Adding dequantize_val and requantize_val"


Differential Revision: [D16678276](https://our.internmc.facebook.com/intern/diff/D16678276)

```
$> ./build/bin/quantized_test
Running main() from ../third_party/googletest/googletest/src/gtest_main.cc
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from TestQTensor
[ RUN      ] TestQTensor.QuantDequantAPIs
[       OK ] TestQTensor.QuantDequantAPIs (0 ms)
[ RUN      ] TestQTensor.RoundingMode
[       OK ] TestQTensor.RoundingMode (0 ms)
[ RUN      ] TestQTensor.Item
[       OK ] TestQTensor.Item (0 ms)
[ RUN      ] TestQTensor.EmptyQuantized
[       OK ] TestQTensor.EmptyQuantized (0 ms)
[----------] 4 tests from TestQTensor (0 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (0 ms total)
[  PASSED  ] 4 tests.
```
